design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,777 @@
+the creation of the following roles:
+
+* `crossplane:composite:composites.example.org:aggregate-to-edit`
+* `crossplane:composite:composites.example.org:aggregate-to-view`
+
+The `aggregate-to-edit` role aggregates to the `crossplane`, `crossplane-admin`,
+and `crossplane-edit` roles, as well as any namespace-aligned `edit` role for a
+namespace in which the composite resource may claimed. The `aggregate-to-view`
+cluster role aggregates to the `crossplane-view` cluster role, and any `view`
+role for a namespace in which the composite resource may be claimed.
+
+Composite resources may be claimed in any namespace that has an annotation with
+the key `rbac.crossplane.io/composites.example.org`, and the value `xrd-enabled`
+(where `composites.example.org` is the name of the XRD). When the RBAC manager
+encounters a namespace with one or more such annotations, it creates `edit` and
+`view` cluster roles for that namespace.
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: example
+  annotations:
+    rbac.crossplane.io/composites.example.org: xrd-enabled
+    rbac.crossplane.io/examplecomposites.xr.example.org: xrd-enabled
+```
+
+The above namespace, for example, will result in the creation of the following
+cluster role. Note that the latter two role selectors are derived from the
+annotations of the namespace.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane-ns-example-edit
+aggregationRule:
+  clusterRoleSelectors:
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-ns-edit: "true"
+      rbac.crossplane.io/base-of-ns-edit: "true"
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-ns-edit: "true"
+      rbac.crossplane.io/xrd: composites.example.org
+  - matchLabels:
+      rbac.crossplane.io/aggregate-to-ns-edit: "true"
+      rbac.crossplane.io/xrd: examplecomposites.xr.example.org
+```
+
+Below is an example of an `edit` composite resource cluster role.
+
+```yaml
+# This composite resource role aggregates up to any namespace aligned cluster
+# role that declares compatibility by including the XRD label. The same role
+# also aggregates to the crossplane-edit role (and thus transitively to the
+# crossplane-admin role).
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:composite:examplecomposites.xr.example.org:aggregate-to-edit
+  labels:
+    rbac.crossplane.io/aggregate-to-edit: "true"
+    rbac.crossplane.io/aggregate-to-ns-edit: "true"
+    rbac.crossplane.io/xrd: examplecomposites.xr.example.org
+rules:
+- apiGroups: [xr.example.org]
+  resources:
+  - exampleclaims
+  - exampleclaims/status
+  # Composite resources are cluster scoped. Binding this cluster role at the
+  # namespace scope will not actually grant access to these composite resources.
+  # Including them in this cluster role allows one crossplane:composite role to
+  # aggregate to both the crossplane-ns-example-edit and crossplane-edit roles.
```

Comment:
This would be deferred to a Crossplane user to build right?

Reply:
Sorry, I'm not quite sure what you're asking.